kf_model_fhir/mappers/loader.py

```diff
@@ -0,0 +1,93 @@
+"""
+Export the following environmental variables first: KF_WAREHOUSE_DB_URL, FHIR_USER, FHIR_PASS
+For the detailed explanaion of arguments, please execute python loader.py -h.
+"""
+import os
+import argparse
+from urllib.parse import urlparse
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from pprint import pformat, pprint
+from pathlib import Path
+import json
+
+import sqlalchemy as sa
+
+from kf_model_fhir.client import FhirApiClient
+from resources.kfdrc_patient import yield_kfdrc_patients
+
+
+def db_study_url(db_maintenance_url, study_id):
+    return urlparse(db_maintenance_url)._replace(path=f"/{study_id}").geturl()
+
+
+def send_resource(payload):
+    endpoint = f'{client.base_url}/{payload["resourceType"]}/{payload["id"]}'
+    success, result = client.send_request('PUT', endpoint, json=payload)
+    return success, result, payload
+
+
+def consume_futures(futures):
+    for future in as_completed(futures):
+        success, result, payload = future.result()
+        print(f'Sent {payload["id"]}')
+        if not success:
+            raise Exception(
+                f'''
+                Failed to submit:\n{pformat(payload)}\n{'-'*80}\n{pformat(result)}
+                '''
+            )
+
+
+# Read in evironmental variables
+KF_WAREHOUSE_DB_URL = os.getenv('KF_WAREHOUSE_DB_URL')
+FHIR_USER = os.getenv('FHIR_USER')
+FHIR_PASS = os.getenv('FHIR_PASS')
+
+# Read in CLI arguments
+parser = argparse.ArgumentParser()
+parser.add_argument('study_id', help='Kids First study ID (e.g. SD_PREASA7S)')
+parser.add_argument(
+    '-t', '--target_url',
+    default='http://localhost:8000',
+    help='Target service URL where data will be loaded into'
+)
+parser.add_argument(
+    '--no_cache',
+    action='store_true',
+    help='A flag to skip caching payloads to a local file'
+)
+
+# Parse arguments
+args = parser.parse_args()
+study_id = args.study_id
+target_url = args.target_url
+no_cache = args.no_cache
+if not no_cache:
+    cache_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'cache')
+    Path(cache_dir).mkdir(parents=True, exist_ok=True) 
+
+# Create DB engine connection
+engine = sa.create_engine(
+    db_study_url(KF_WAREHOUSE_DB_URL, study_id),
+    connect_args={'connect_timeout': 5},
+    server_side_cursors=True
+)
+schema='Ingest:initial-ingest:GuidedTransformStage'
+table = f'"{schema}".default'
+
+# Intantiate FHIR API client
+client = FhirApiClient(base_url=target_url, auth=(FHIR_USER, FHIR_PASS))
+
+with ThreadPoolExecutor(max_workers=10) as tpex:
```

Comment:
Again, I know its a prototype but is it expected that the loader should hard code the send patient loop? Will Avi also append his code for the send condition loop? Maybe the sending logic should be in a method or something:

```python
def send_entities(yield_entity_method):
    # Send entities
     pass

senders = [yield_kfdrc_patients, yield_kfdrc_conditions] 
for sender in senders:
    send_entities(sender)
```  
Then Avi or whoever would just add their send method to the list?

Reply:
@znatty22, I am trying to come up with a decent way of doing the above. Other `yield_resources` functions than `yield_kfdrc_patients` require patient resource IDs created [here](https://github.com/kids-first/kf-model-fhir/blob/participant-mapper/kf_model_fhir/mappers/resources/kfdrc_patient.py#L134), so that they can reference patients at `subject`. We can created a dictionary to store patient IDs as follows:

```python
kfdrc_patients = {}
builders = [
    yield_kfdrc_patients(engine, table, study_id),
    yield_kfdrc_conditions(engine, table, study_id, kfdrc_patients)
] 

for builder in builders:
    send_resources(builder)
```
`kfdrc_patients` can be populated during `send_resources(builder)` for `yield_kfdrc_patients(engine, table, study_id)`. But, since `kfdrc_patients` for `yield_kfdrc_conditions(engine, table, study_id, kfdrc_patients)` was already consumed as `{}`, the latter builder cannot use the populated `kfdrc_patients`.

Reply:
https://github.com/kids-first/kf-model-fhir/blob/participant-mapper/kf_model_fhir/mappers/loader.py#L78-L91

I wanted to implement the above, but it seems not simple for other resources than patients require a patient cache as subject reference.

Reply:
> is it expected

Yep

> Will Avi also append his code for the send condition loop? Maybe 

What value exists in developing this further?